### PR TITLE
refactor: split parser info into separate files

### PR DESCRIPTION
**Pull request by lewis6991**
~~- Also fix `auto_install`.~~

### CODE REVIEW

**Comment by clason** [comment]
Maybe we can rewrite `get_available` to not needing to require every single parser? (Or is there a convenient metatable method for that? Maybe `__pairs`?)

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -28,13 +28,25 @@ M.configs = setmetatable({}, {
   end
 })
 
+--- @return string[]
+local function get_parser_names()
+  local path = vim.fs.joinpath(vim.fn.fnamemodify(debug.getinfo(1, 'S').source:sub(2), ':p:h'), 'parsers')
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:33]: `get_package_path` could be moved here instead (since `install` imports `parsers` anyway?)

**Comment by lewis6991** [comment]
> Maybe we can rewrite `get_available` to not needing to require every single parser? (Or is there a convenient metatable method for that? Maybe `__pairs`?)
> 
> It doesn't matter in the scripts, but maybe the two places in `install` that iterate over the (sorted) `parser_list` may profit? (Could have an `__ipairs` method that just iterates over the sorted list of names? Or is that too magic?)

Not possible with Lua 5.1 😞

**Comment by clason** [comment]
@lewis6991 can you pull out the auto_install fix into a separate PR (or just commit to `main`)?

Note to self: I missed the (re-)introduction of `util.lua`; should be used in the scripts as well.

File: lua/nvim-treesitter/install.lua [hunk 2]
```diff
@@ -243,6 +243,7 @@ local function do_download_tar(logger, repo, project_name, cache_dir, revision,
 
   logger:debug('Creating temporary directory: ' .. temp_dir)
   --TODO(clason): use vim.fn.mkdir(temp_dir, 'p') in case stdpath('cache') is not created
+  util.delete(temp_dir)
```
Comment (clason) [inline lua/nvim-treesitter/install.lua:246]: Aren't you already deleting that dir in line 221?

**Comment by clason** [comment]
Two open issues with the refactor:

1. This breaks the `update-readme` script.
2. This breaks the documented way of adding new parsers in the README.

Issue 1 is a bit of a blocker (but can be worked around, I'm sure). Now that nvim-treesitter no longer handles modules, is 2. even something we want to keep doing? It's not clear to me what benefit there is for having nvim-treesitter manage local parsers (as the main benefit -- the lockfile update -- doesn't apply to user-added parsers).

File: lua/nvim-treesitter/install.lua (hunk 2 above)
Comment (lewis6991) [reply lua/nvim-treesitter/install.lua:246]: Yeah, I didn't understand the issue when I added this.

**Comment by clason** [comment]
@lewis6991 The scripts work now (and have become simpler, which makes me broadly in favor of this change).

Could you re-create the parser files (maybe after the D parser change to avoid another rebase)?

One issue is that lua-language-server now complains about
> The same file is required with different names.

on every `require('nvim-treesitter.parsers')`.

**Comment by clason** [comment]
I just did a short test of this branch, and it significantly increases startuptime for me (3ms -> 23ms) with
```lua
ensure_install = { 'core', 'julia', 'markdown', 'markdown_inline' } 
```

**Comment by clason** [comment]
Maybe. The issue is that `get_available` needs to loop through all parser configs to check the tiers; maybe there's a smarter way for this. Otherwise it's indeed looking like a bad trade-off -- unless we want to drop tiers from the installation entirely, or replace (some of) them with hard-coded lists.

**Comment by clason** [comment]
I think I'll close this for now, to reduce the noise from my force-pushes. We can pick this back up and reopen when someone has a good idea for handling parser tiers outside of the individual parser configs.

**Pull request closed without merging**